bdns/servers.go

```diff
@@ -0,0 +1,137 @@
+package bdns
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+// serverProvider represents a type which can provide a list of addresses for
+// the bdns to use as DNS resolvers. Different implementations may provide
+// different strategies for providing addresses, and may provide different kinds
+// of addresses (e.g. host:port combos vs IP addresses).
+type ServerProvider interface {
+	Addrs() ([]string, error)
+	Stop()
+}
+
+// staticProvider stores a list of host:port combos, and provides that whole
+// list in randomized order when asked for addresses. This replicates the old
+// behavior of the bdns.impl's servers field.
+type staticProvider struct {
+	servers []string
+}
+
+var _ ServerProvider = &staticProvider{}
+
+func NewStaticProvider(servers []string) *staticProvider {
+	return &staticProvider{servers: servers}
+}
+
+func (sp *staticProvider) Addrs() ([]string, error) {
+	if sp.servers == nil || len(sp.servers) == 0 {
+		return nil, fmt.Errorf("no servers configured")
+	}
+	r := make([]string, len(sp.servers))
+	perm := rand.Perm(len(sp.servers))
+	for i, v := range perm {
+		r[i] = sp.servers[v]
+	}
+	return r, nil
+}
+
+func (sp *staticProvider) Stop() {}
+
+// dynamicProvider uses DNS to look up the set of IP addresses which correspond
+// to its single host. It returns this list in random order when asked for
+// addresses, and refreshes it regularly using a goroutine started by its
+// constructor.
+type dynamicProvider struct {
+	host   string
+	addrs  map[string][]uint16
+	mut    sync.RWMutex
+	cancel chan interface{}
+}
+
+var _ ServerProvider = &dynamicProvider{}
+
+// StartDynamicProvider constructs a new dynamicProvider and starts its
+// auto-update goroutine. The auto-update process queries DNS for SRV records
+// at refresh intervals and uses the resulting IP/port combos to populate the
+// list returned by Addrs. The update process ignores the Priority and Weight
+// attributes of the SRV records.
+func StartDynamicProvider(server string, refresh time.Duration) (*dynamicProvider, error) {
+	if server == "" {
+
+	}
+	dp := dynamicProvider{
+		host:   server,
+		addrs:  make(map[string][]uint16),
+		mut:    sync.RWMutex{},
+		cancel: make(chan interface{}),
+	}
+	err := dp.update()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start dynamic provider: %w", err)
+	}
```

Comment:
A subtle thing about this: Most of our boulder components don't do potentially blocking/erroring network work during startup. That avoids creating hard ordering dependencies for startup in prod (even though we do observe ordering in dev to minimize spurious errors).

Since this is going to a system service (DNS) that is expected to always be up at the time Boulder starts, it's maybe not a big deal. But I wouldn't be surprised to learn that this creates some deployment pain. The alternative would be for all queries to block on our first successful update (or time out if that takes too long).

Reply:
Yeah, I went back and forth on this a couple times in my head. I didn't want to have the first queries block, because that introduces weird conditional waiting/polling logic in the `Addrs` method, which is currently very straightforward. That's also why I named it `StartFoo` instead of `New...` or `Make...`, to make it clear that it is doing Actual Work. But I can definitely add some logic to `Addrs` instead, if you'd prefer.

Reply:
Let's continue with it as-is and just keep this in mind as something that might want to change later.